Personal copilot blog (#1413)
huggingface/blog: personal_copilot.md

---
title: "Personal Copilot: Train Your Own Coding Assistant" 
thumbnail: /blog/assets/159_safecoder/thumbnail.jpg
authors:
- user: smangrul
- user: sayakpaul
---

# Personal Copilot: Train Your Own Coding Assistant

> ## Introduction

Reviewer: I think it's okay to skip this header.

Scraping code contents from GitHub is straightforward with the [Python GitHub API](https://github.com/PyGithub/PyGithub). However, depending on the number of repositories and the number of code files within a repository, one might easily run into API rate-limiting issues. 

To prevent such problems, we decided to clone all the public repositories locally and extract the contents from them instead of through the API. We used the `multiprocessing` module from Python to download all repos in parallel, as shown in [this download script](https://github.com/sayakpaul/hf-codegen/blob/main/data/parallel_clone_repos.py). 

A repository can often contain non-code files such as images, presentations and other assets. We’re not interested in scraping them. We created a [list of extensions](https://github.com/sayakpaul/hf-codegen/blob/f659eba76f07e622873211e5b975168b634e6c22/data/prepare_dataset.py#L17C1-L49C68) to filter them out. To parse code files other than Jupyter Notebooks, we simply used the "utf-8" encoding. For notebooks, we only considered the code cells.

We also excluded all file paths that were not directly related to code. These include: `.git`, `__pycache__`, and `xcodeproj`. 

> To keep the serialization of this content relatively memory-friendly, we used chunking and the feather format. Refer to [this script](https://github.com/sayakpaul/hf-codegen/blob/main/data/prepare_dataset.py) for the full implementation. 

Reviewer: ```suggestion
To keep the serialization of this content relatively memory-friendly, we used chunking and the [feather format](https://arrow.apache.org/docs/python/feather.html#:~:text=Feather%20is%20a%20portable%20file,Python%20(pandas)%20and%20R.). Refer to [this script](https://github.com/sayakpaul/hf-codegen/blob/main/data/prepare_dataset.py) for the full implementation. 
```

> [The final dataset is available in the Hub](https://huggingface.co/datasets/sayakpaul/hf-codegen-v2), and it looks like this:

Reviewer: ```suggestion
The final dataset is [available on the Hub](https://huggingface.co/datasets/sayakpaul/hf-codegen-v2), and it looks like this:
```

![hf-stack-full](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/personal_copilot/hf-stack-full.png)

For this blog, we considered the top 10 Hugging Face public repositories, based on stargazers. They are the following: 

> ['transformers', 'pytorch-image-models', 'datasets', 'diffusers', 'peft', 'tokenizers', 'accelerate', 'text-generation-inference', 'chat-ui', 'deep-rl-class']

[This is the code we used to generate this dataset](https://github.com/pacman100/DHS-LLM-Workshop/tree/main/personal_copilot/dataset_generation), and [this is the dataset in the Hub](https://huggingface.co/datasets/smangrul/hf-stack-v1). Here is a snapshot of what it looks like: 
![hf-stack-v1](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/personal_copilot/hf-stack-v1.png)

To reduce the project complexity, we didn’t consider deduplication of the dataset. If you are interested in applying deduplication techniques for a production application, [this blog post](https://huggingface.co/blog/dedup) is an excellent resource about the topic in the context of code LLMs.

## Finetuning your own Personal Co-Pilot 

> In this section, we show how to fine-tune `bigcode/starcoder` (15.5B params), `bigcode/starcoderbase-1b` (1B params), `Deci/DeciCoder-1b` (1B params) on a single A100 40GB Colab Notebook using 🤗 PEFT (Parameter-Efficient Fine-Tuning). Then, we'll show how to fully finetune the `bigcode/starcoder` (15.5B params) on a machine with 8 A100 80GB GPUs using 🤗 Accelerate's FSDP integration. The training objective is fill in the middle (FIM) wherein parts of a training sequence are moved to the end, and the reordered sequence is predicted autoregressively.

Reviewer: ```suggestion
In this section, we show how to fine-tune the following models: [`bigcode/starcoder`](https://hf.co/bigcode/starcoder) (15.5B params), [`bigcode/starcoderbase-1b`](https://hf.co/bigcode/starcoderbase-1b) (1B params), [`Deci/DeciCoder-1b`](https://hf.co/Deci/DeciCoder-1b) (1B params). We'll use a single A100 40GB Colab Notebook using 🤗 PEFT (Parameter-Efficient Fine-Tuning) for all the experiments. Additionally, we'll show how to fully finetune the `bigcode/starcoder` (15.5B params) on a machine with 8 A100 80GB GPUs using 🤗 Accelerate's FSDP integration. The training objective is [fill in the middle (FIM)](https://arxiv.org/abs/2207.14255), wherein parts of a training sequence are moved to the end, and the reordered sequence is predicted auto-regressively.
```

In the above calculations, we didn't consider memory required for intermediate activation checkpointing which is considerably huge. We leverage Flash Attention V2 and Gradient Checkpointing to overcome this issue. 

1. For QLoRA along with flash attention V2 and gradient checkpointing, the total memory occuiped by the model on a single A100 40GB GPU is **26 GB** with a **batch size of 4**.
2. For full fine-tuning using FSDP along wth Flash Attention V2 and Gradient Checkpointing, the memory occupied per GPU ranges between **70 GB to 77.6 GB** with a **per_gpu_batch_size of 1**.

Please refer to the [model-memory-usage](https://huggingface.co/spaces/hf-accelerate/model-memory-usage) to easily calculate how much vRAM is needed to train and perform big model inference on a model hosted on the 🤗 Hugging Face Hub.

## Full Finetuning

> We will look at how to do full fine-tuning of starcoder-15B on 8 A100 80GB GPUs using PyTorch Fully Sharded Data Parallel (FSDP) technique. For more information on FSDP, please refer [Fine-tuning Llama 2 70B using PyTorch FSDP](https://huggingface.co/blog/ram-efficient-pytorch-fsdp) and [Accelerate Large Model Training using PyTorch Fully Sharded Data Parallel](https://huggingface.co/blog/pytorch-fsdp).

Reviewer: Let's maintain consistency when referring to model checkpoints. Let's maybe follow `bigcode/starcoder15B`.

## PEFT 

We will look at how to use QLoRA for fine-tuning starcoder-15B on a single A100 40GB GPU using 🤗 PEFT. For more information on QLoRA and PEFT methods, please refer [Making LLMs even more accessible with bitsandbytes, 4-bit quantization and QLoRA](https://huggingface.co/blog/4bit-transformers-bitsandbytes) and [🤗 PEFT: Parameter-Efficient Fine-Tuning of Billion-Scale Models on Low-Resource Hardware](https://huggingface.co/blog/peft).

**Resources**
1. Codebase: [link](https://github.com/pacman100/DHS-LLM-Workshop/tree/main/personal_copilot/training). It uses the recently added Flash Attention V2 support in Transformers. 
2. Colab notebook : [link](https://colab.research.google.com/drive/1Tz9KKgacppA4S6H4eo_sw43qEaC9lFLs?usp=sharing). Make sure to choose A100 GPU with High RAM setting.
3. Model: [bigcode/stacoder](https://huggingface.co/bigcode/starcoder)
4. Dataset: [smangrul/hf-stack-v1](https://huggingface.co/datasets/smangrul/hf-stack-v1)
5. Trained Model: [smangrul/peft-lora-starcoder15B-v2-personal-copilot-A100-40GB-colab](https://huggingface.co/smangrul/peft-lora-starcoder15B-v2-personal-copilot-A100-40GB-colab)

> The command to launch training is given at [run_peft.sh](https://github.com/pacman100/DHS-LLM-Workshop/blob/main/personal_copilot/training/run_peft.sh). The total training time was **12.5 Hours**. Taking the cost of **$1.10 / hr** based on [lambdalabs](https://lambdalabs.com/service/gpu-cloud/pricing), the total cost would be **$13.75**. That's pretty good 🚀! In terms of cost, it's **7.8X** lower than the cost for full fine-tuning.  

Reviewer: We have already talked about the memory requirements with and without QLoRA. So, I guess it's okay to skip that part here? You have already done it but ensuring if we don't want to add a sentence about the memory part.

Author: Here, we are comparing the cost of training. I think this is important metrics from the end users point of view.

## Comparison

Below plot shows the eval loss, train loss and learning rate scheduler for QLoRA vs full fine-tuning. We observe that full fine-tuning leads to slightly lower loss and converges a bit faster compared to QLoRA. The learning rate for peft fine-tuning is 10X more than that of full fine-tuning.

![plots](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/personal_copilot/full_finetuning_vs_qlora.png)

To make sure that our PEFT model doesn't lead to catastrophic forgetting, we run the Python Human Eval on it. Below are the results. We can observe that the performance on humaneval-python is comparable for both the base `StarCoder-15B` and the PEFT model `smangrul/peft-lora-starcoder15B-v2-personal-copilot-A100-40GB-colab`.

| | |
|---|---|

> | Model | Pass@1 |

Reviewer: What does `Pass@1` denote?

|StarCoder-15B | 33.57|
|smangrul/peft-lora-starcoder15B-v2-personal-copilot-A100-40GB-colab| 33.37 |

As we don't have a benchmark, we will look at some qualitative samples. Inference Code with various examples for full fine-tuned model and peft model are available at [Full_Finetuned_StarCoder_Inference.ipynb](https://github.com/pacman100/DHS-LLM-Workshop/blob/main/personal_copilot/inference/Full_Finetuned_StarCoder_Inference.ipynb) and [PEFT_StarCoder_Inference.ipynb](https://github.com/pacman100/DHS-LLM-Workshop/blob/main/personal_copilot/inference/PEFT_StarCoder_Inference.ipynb), respectively. In our manual analysis, we noticed that the QLoRA led to slight overfitting and as such we down weigh it by creating new weighted adapter with weight 0.8 via `add_weighted_adapter` utility of PEFT.

We will look at 2 code infilling examples wherein the task of the model is to fill the part denoted by placeholder `<FILL_ME>`. We will consider infliing completions from Github Copilot, QLoRA fine-tuned model and full fine-tuned model. 

> ![qualitative_comparison_1](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/personal_copilot/qlora_vs_finetune_1.png)
> Qualitative Example 1

Reviewer: https://huggingface.co/blog/stackllama shows a good way of captioning the figures:

```suggestion
![qualitative_comparison_1](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/personal_copilot/qlora_vs_finetune_1.png)
*Qualitative Example 1*
```

![ie_1](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/personal_copilot/inference_endpoint_1.png)
![ie_2](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/personal_copilot/inference_endpoint_2.png)

### Setting up the VS Code Extension
Follow the installation steps mentioned [here](https://github.com/huggingface/llm-vscode#installation). Replace the endpoint via the settings to the HF Inference endpoint in the field highlighted below.

![vs_code_endpoint](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/personal_copilot/vs_code_endpoint.png)

Usage will look like below:

![code_completion](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/personal_copilot/vs_code_completion_usage.png)

# Finetuning your own Code Chat Assistant

So far, the models we trained were specifically trained as personal co-pilot for code completion tasks. They aren't trained to carry out conversations or for question answering. `Octocoder` and `StarChat` are great examples of such models. This section briefly describes how to achieve that.

> Resources: 

Reviewer: ```suggestion
**Resources**
```

In the interest of consistency.

---
title: "HugCoder 🤗: Train Your Own Coding Assistant 🚀" 

> thumbnail: /blog/assets/159_safecoder/thumbnail.jpg

Reviewer: To be updated. And an entry needs to be added to "_blog.yml".

In the ever-evolving landscape of programming and software development, the quest for efficiency and productivity has led to remarkable innovations. One such innovation is the emergence of code generation models such as [Codex](https://openai.com/blog/openai-codex). These models have demonstrated remarkable capabilities in generating human-like code snippets, thereby showing immense potential as coding assistants.

However, while these pre-trained models can perform impressively across a range of tasks, there's an exciting possibility lying just beyond the horizon: the ability to tailor a code generation model to your specific needs. Think of personalized coding assistants which could be leveraged at an enterprise scale. 

In this blog post, HugCoder 🤗, a code LLM, fine-tuned on the code contents from the public repositories of [huggingface GitHub organization](https://github.com/huggingface). We will discuss our data collection workflow, our training experiments, and some interesting results. We will leave you with a couple of further extensions of this project for experimentation. 

Let’s begin 🚀

[Insert a Space link here to let folks try it out immediately?]

[Insert ToC]

> https://github.com/pacman100/blog/assets/13534540/f792b506-c31a-4f73-a321-3333902c3c52

Reviewer: I think it's better the video out from here and instead either do a Space or a shorter GIF.

Author: Hello Sayak, code completion task can only be done in IDE and as such Space is not suitable for it. GIF is too short to show real-world usage because it hides the latency, the iterative process of accepting generations, writing comments/code, updating the generations and so on. As such, I feel this video is the best way to showcase the abilities of the Personal Co-pilot else people won't be interested in the blog.

Reviewer: Makes sense. Maybe give a little teaser text to supplement the video so that it excites the reader to check it out.

Our dataset prepared this way is available [here](https://huggingface.co/datasets/sayakpaul/hf-codegen-v2) and it looks like so:

![hf-stack-full](assets/170_personal_copilot/hf-stack-full.png)

For this blog, we consider the top 10 Hugging Face public repositories based on stargazers. They are the following: 

> ['transformers', 'pytorch-image-models', 'datasets', 'diffusers', 'peft', 'tokenizers', 'accelerate', 'text-generation-inference', 'chat-ui', 'deep-rl-class']

The code for this dataset generation is [here](https://github.com/pacman100/DHS-LLM-Workshop/tree/main/personal_copilot/dataset_generation), and the dataset can be found [here](https://huggingface.co/datasets/smangrul/hf-stack-v1). Here is a snapshot of the dataset: 
![hf-stack-v1](assets/170_personal_copilot/hf-stack-v1.png)

In the interest of less complexity, we didn’t consider deduplication of the dataset. But for production-ready applications, deduplication should be considered an inseparable step of the data collection pipeline. We welcome you to check out [this blog post](https://huggingface.co/blog/dedup) if you’re interested to learn more about this topic in the context of code LLMs. 

> # Finetuning your own Personal Co-Pilot 

Reviewer: We usually keep the section headers to "##".

> In this section, we will show you how to fine-tune `bigcode/starcoder` (15.5B params), `bigcode/starcoderbase-1b` (1B params), `Deci/DeciCoder-1b` (1B params) on a single A100 40GB Colab Notebook using 🤗 PEFT. Then, we will show you how to fully finetune the `bigcode/starcoder` (15.5B params) on a machine with 8 A100 80GB GPUs using 🤗 Accelerate's FSDP integration.

Reviewer: * Give one/two lines introduction about the existing code LLMs available on the Hub. 
* Link the models. 
* Link PEFT.

Why PEFT? Full Fine-tuning is expensive. Let’s put some numbers to put things in perspective:

> Minimum GPU memory required for Full Fine-tuning:

Reviewer: ```suggestion
Minimum GPU memory required for full fine-tuning:
```

1. Weight: 2 bytes (Mixed-Precision training)
2. Weight gradient: 2 bytes
3. Optimizer state when using Adam: 4 bytes for original FP32 weight + 8 bytes for first and second moment estimates
4. Cost per parameter adding all the above: 16 bytes per parameter 
5. **15.5B model -> 248GB of GPU memory without even considering huge memory requirements for storing intermediate activations -> minimum 4X A100 80GB GPUs required**

Minimum GPU memory required for QLoRA method:
Using 

> > trainable params: 110,428,160 || all params: 15,627,884,544 || trainable%: 0.7066097761926236

Reviewer: I would structure this section as follows:

Introduce a line saying something like: Since the hardware requirements are huge, we'll be using parameter-efficient fine-tuning using QLoRA (link). Here are the minimal GPU memory requirements for fine-tuning StarCoder using QLoRA

Reviewer 2: Agreed :) (Lysandre and I are working towards also putting it in more docs :) )

1. Base model Weight: 0.5 bytes * 15.51B frozen params  = 7.755 GB
2. Adapter weight: 2 bytes * 0.11B trainable params        = 0.22GB
3. Weight gradient: 2 bytes * 0.11B trainable params       = 0.12GB
4. Optimizer state when using Adam:  4 bytes * 0.11B trainable params * 3 = 1.32GB

> 5. **Adding all of the above -> 9.51 GB ~10GB -> 1 A100 40GB GPU required** 🤯

Reviewer: It's only 10 GB no? They why an A100 is needed?

Author: Because of intermediate activations for a long sequence length of 2048 and a batch size of 4 for proper training + Flash Attention V2 support for A100 is very efficient. As you can see at the end of the section, you can get the GPU memory required is 26GB which can be accommodated on A100 40GB GPU.

Reviewer: Let's clarify that right away. I am sure the reader would have the same question at this point.

**Computing the cost for intemediate Activations**:

1. Sequence length s = 2048
2. Number of layers n = 40
3. Number of attention heads a = 48
4. Hidden dimension h = 6144
5. Batch size b = 4

> 8. LoRA rank r = 32

Reviewer: This seems irrelevant for the full activation memory requirement computation. Can't we shift it to the section below?

Author: Below, we also compute LoRA activations memory for which we require rank r

Reviewer: Then, it makes sense to introduce that ad-hoc. Otherwise, it is still irrelevant IMO.

7. Refer paper Reducing [Activation Recomputation in Large Transformer Models](https://arxiv.org/pdf/2205.05198.pdf) for approximate derivation of memory occupied by activations. Below is the screenshot of the relevant snippet showing activation memory per layer

> ![activation memory per layer](assets/170_personal_copilot/activation_memory_computation.png)

Reviewer: This section seems a bit incoherent. Maybe before "**Computing the cost for intemediate Activations**:", it would make sense to write a line about it discussing the need to know about the cost of computing intermediate activations?

Author: Thinking of removing this part as it might be more involved for this blog post.

> The above formula considers half-precision. As starcoder uses MQA, it will result in the following equation:

Reviewer: What is MQA?

Author: Multi Query Attention

Reviewer: Reader might not know it.

```
Activations memory per layer = sbh(30+4/a+5as/h)
Total Activations memory = (2048*4*6144(30+(4/48)+((5*48*2048)/6144)) * 40)/(1024^3) GB = 206 GB
```

> For LoRA, we need to account for the activations of the LoRA layers based on the targets `c_proj,c_attn,q_attn,c_fc,c_proj`:

Reviewer: We're using LoRA and QLoRA. Readers might get confused about that.

Reviewer: ```suggestion
When using LoRA, we only need to account for the activations of the LoRA layers based on the target modules we're adapting (`c_proj, c_attn, q_attn, c_fc, c_proj`). So the requirements drastically reduces 🤯
```

Author: No, no. LoRA activations are on top of the base model activations. It should be `also`

Reviewer: Feel free to modify.

```
LoRA Activations memory per layer = sbh(18+2/a+6r/h)
Total LoRA Activations memory = (2048*4*6144(18+(2/48)+((6*32)/6144)) * 40)/(1024^3) = 34 GB
```

> To overcome this, we leverage Flash Attention V2 and Gradient Checkpointing. 

Reviewer: What are we overcoming here?

Author: the huge cost of storing intermediate activation memory

Reviewer: That's not clear from the reading. My questions are from the perspective of a reader.

1. For QLoRA along with Flash Attention V2 and Gradient Checkpointing, the total memory occuiped by the model on a single A100 40GB GPU is **26 GB** with a **batch size of 4**.

> 2. For Full Finetuning using FSDP along wth Flash Attention V2 and Gradient Checkpointing, the memory occupied per GPU ranges between **70 GB to 77.6 GB** with a **per_gpu_batch_size of 1**.

Reviewer: Let's maintain:

* flash attention V2
* gradient checkpointing
* full fine-tuning

> 

Reviewer: Maybe discuss the training objective as well? It's not immediately noticeable that FIM was used.

## PEFT 

> Resources: 

Reviewer 3: Here I'm missing an introductory sentence. As a reader, it feels like something new starts, so having a sentence that makes the transition from the previous section clear would help. Same applies to the next section "Full Finetuning".

Also, I wonder: Would it be better to start with "Full Finetuning" and then show "PEFT". Then the PEFT numbers may look even more impressive.

5. Trained Model: [smangrul/peft-lora-starcoder15B-v2-personal-copilot-A100-40GB-colab](https://huggingface.co/smangrul/peft-lora-starcoder15B-v2-personal-copilot-A100-40GB-colab)

The command to launch training is given below:
```
python train.py \
    --model_path "bigcode/starcoder" \
    --dataset_name "smangrul/hf-stack-v1" \
    --subset "data" \
    --data_column "content" \
    --split "train" \
    --seq_length 2048 \
    --max_steps 2000 \
    --batch_size 4 \
    --gradient_accumulation_steps 4 \
    --learning_rate 5e-4 \
    --lr_scheduler_type "cosine" \
    --weight_decay 0.01 \
    --num_warmup_steps 30 \
    --eval_freq 100 \
    --save_freq 100 \
    --log_freq 25 \
    --num_workers 4 \
    --bf16 \
    --no_fp16 \
    --output_dir "peft-lora-starcoder15B-v2-personal-copilot-A100-40GB-colab" \
    --fim_rate 0.5 \
    --fim_spm_rate 0.5 \
    --use_flash_attn \
    --use_peft_lora \
    --lora_r 32 \
    --lora_alpha 64 \
    --lora_dropout 0.0 \
    --lora_target_modules "c_proj,c_attn,q_attn,c_fc,c_proj" \
    --use_4bit_qunatization \
    --use_nested_quant \
    --bnb_4bit_compute_dtype "bfloat16"
```

> The total training time was **12.5 Hours**. Taking the cost of **$1.10 / hr** based on [lambdalabs](https://lambdalabs.com/service/gpu-cloud/pricing), the total cost would be **$13.75**. That's pretty good! 🚀

Reviewer: ```
That's pretty good! 🚀
```

Let's leave this to the readers to judge ;)

Author: Hmm, it is like mentioning in the paper that we got good results and that too while being efficient.

## Full Finetuning

> 

Reviewer 4: As before, would be nice to have a sentence giving a small intro here.

Author: Done

> 1. Codebase: [link](https://github.com/pacman100/DHS-LLM-Workshop/tree/main/personal_copilot/training). It uses monkey patch of Flash Attention V2. 

Reviewer: Readers might not know the monkey-patching bit you're talking about. Let's link "monkey patch of flash attention V2".

```
accelerate launch --config_file "configs/fsdp_config.yaml"  train.py \
    --model_path "bigcode/starcoder" \
    --dataset_name "smangrul/hf-stack-v1" \
    --subset "data" \
    --data_column "content" \
    --split "train" \
    --seq_length 2048 \
    --max_steps 2000 \
    --batch_size 1 \
    --gradient_accumulation_steps 2 \
    --learning_rate 5e-5 \
    --lr_scheduler_type "cosine" \
    --weight_decay 0.01 \
    --num_warmup_steps 30 \
    --eval_freq 100 \
    --save_freq 500 \
    --log_freq 25 \
    --num_workers 4 \
    --bf16 \
    --no_fp16 \
    --output_dir "starcoder-personal-copilot-A100-40GB-colab" \
    --fim_rate 0.5 \
    --fim_spm_rate 0.5 \
    --use_flash_attn
```

The total training time was **9 Hours**. Taking the cost of $12.00 / hr based on [lambdalabs](https://lambdalabs.com/service/gpu-cloud/pricing) for 8x A100 80GB GPUs, the total cost would be **$108**. That's **7.8X** higher than the cost for training with QLoRA. 

## Comparison

Below plot shows the eval loss, train loss and learning rate scheduler for QLoRA vs Full Finetuning.

![plots](assets/170_personal_copilot/full_finetuning_vs_qlora.png)

> As we don't have a benchmark, we will look at some qualitative samples.

Reviewer 3: A sentence or two on what we learn from these graphs would be nice.

Inference Code for Full Fine-tuned model:

```python
import os
os.environ["CUDA_VISIBLE_DEVICES"]="0"

from dataclasses import dataclass, field
from typing import Optional
import contextlib

import torch
from datasets import load_dataset
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    BitsAndBytesConfig,
    HfArgumentParser,
    AutoTokenizer,
    TrainingArguments,
)
model = "smangrul/starcoder-personal-copilot"
tokenizer = AutoTokenizer.from_pretrained(model, trust_remote_code=True)
model = AutoModelForCausalLM.from_pretrained(
    model, quantization_config=None, 
    device_map=None, 
    trust_remote_code=True, 
    torch_dtype=torch.bfloat16,
)

if not hasattr(model, "hf_device_map"):
    model.cuda()

def get_code_completion(prefix, suffix):
    text = prompt = f"""<fim_prefix>{prefix}<fim_suffix>{suffix}<fim_middle>"""
    model.eval()

>     outputs = model.generate(input_ids=tokenizer(text, return_tensors="pt").input_ids.cuda(), 

Reviewer 3: Maybe use hanging indents like everywhere else?

Reviewer 4: Should be formatted using black as the rest of the HF code-base.

    BitsAndBytesConfig,
    HfArgumentParser,
    AutoTokenizer,
    TrainingArguments,
)
from peft import (
    prepare_model_for_kbit_training,
    LoraConfig,
    get_peft_model,
    PeftModel
)
model = "bigcode/starcoder"
tokenizer = AutoTokenizer.from_pretrained(model, trust_remote_code=True)
model = AutoModelForCausalLM.from_pretrained(
    model, quantization_config=None, 
    device_map=None, 
    trust_remote_code=True, 
    torch_dtype=torch.bfloat16,
)

# model = model.merge_and_unload()
if not hasattr(model, "hf_device_map"):
    model.cuda()

model_id = "smangrul/peft-lora-starcoder15B-v2-personal-copilot-A100-40GB-colab"
model = PeftModel.from_pretrained(model, model_id, adapter_name="personal_copilot")
model.add_weighted_adapter(["personal_copilot"], [0.8], "best_personal_copilot")
model.set_adapter("best_personal_copilot")

# get_code_completion same as above
```

Below screenshots of a table show the predictions by Fully fine-tuned model in comparison with the PEFT QloRA model:
![qualitative_comparison_1](assets/170_personal_copilot/qualitative_comparison_1.png)
![qualitative_comparison_2](assets/170_personal_copilot/qualitative_comparison_2.png)

> We can observe that the generations from both the variants are as per expectations. Awesome! 🚀

Reviewer 3: Would it make sense to show a comparison with GH copilot?

## How do I use it in VS Code?

🤗 VS Code Extension [HF Code Autocomplete](https://marketplace.visualstudio.com/items?itemName=HuggingFace.huggingface-vscode) coupled with hosting the model via [🤗 Inference EndPoints](https://ui.endpoints.huggingface.co/). 

### Setting an inference Inference Endpoint
Below are the screenshots of the Inference Endpoint setting.
![ie_1](assets/170_personal_copilot/inference_endpoint_1.png)
![ie_2](assets/170_personal_copilot/inference_endpoint_2.png)

### Setting up the VS Code Extension
Follow the installation steps mentioned [here](https://github.com/huggingface/huggingface-vscode#installing). Replace the endpoint via the settings to the HF Inference endpoint in the field highlighted below.

![vs_code_endpoint](assets/170_personal_copilot/vs_code_endpoint.png)

Usage will look like below:

> ![code_completion](assets/170_personal_copilot/vs_code_completion_usage.png)

Reviewer: I think it would make sense to 

* include these steps in the GitHub repo. 
* just mention that we can go ahead and use this inside VS Code as an extension with a screenshot. 

This will help us reduce content improving brevity.

Resources: 

1. Codebase: [link](https://github.com/pacman100/DHS-LLM-Workshop/tree/main/code_assistant/training). It uses monkey patch of Flash Attention V2. 
**Note:** Flash V2 support implemented here ignores padding/attention_mask/custom_mask. It is meant for continued pre-training/fine-tuning with dense packing inputs to consume the entire sequence lengths.
2. Colab notebook : [link](https://colab.research.google.com/drive/1XFyePK-3IoyX81RM94JO73CcIZtAU4i4?usp=sharing). Make sure to choose A100 GPU with High RAM setting.
3. Model: [bigcode/stacoderplus](https://huggingface.co/bigcode/starcoderplus)
4. Dataset: [smangrul/code-chat-assistant-v1](https://huggingface.co/datasets/smangrul/code-chat-assistant-v1). Mix of `LIMA+GUANACO` with proper formatting in a ready-to-train format.
5. Trained Model: [smangrul/peft-lora-starcoderplus-chat-asst-A100-40GB-colab](https://huggingface.co/smangrul/peft-lora-starcoderplus-chat-asst-A100-40GB-colab) 

# Dance of LoRAs

> If you have dabbled with Stable Diffusion models and LoRAs for making your own Dreambooth models, you might be familiar with the concepts of combining different LoRAs with different weights, using a loRA model with a different base model than the one on which it was trained. In text/code domain, this remains unexplored territory. We carry about experiments in this regard and have observed very promising findings. You Ready? let's go, let's go, let's go!!! 🚀

Reviewer 3: ```suggestion
If you have dabbled with Stable Diffusion models and LoRAs for making your own Dreambooth models, you might be familiar with the concepts of combining different LoRAs with different weights, using a LoRA model with a different base model than the one on which it was trained. In text/code domain, this remains unexplored territory. We carry out experiments in this regard and have observed very promising findings. Are you ready? Let's go, let's go, let's go!!! 🚀
```

The tone of this paragraph is a bit different than the rest of the article (more energetic). Not a big deal, just my impression.

> ## Mix-and-Match LoRAs

Reviewer: What concepts are we mixing and matching? Maybe give a teaser here?

Author: I don't know how to do that without slowly building from disabled adapter, to task-specific adapter and then combining them. Unlike SD, wherein we can show things visually, it might be difficult for readers to comprehend the same seeing code/text.

Reviewer: We're mixing chat and QA IIRC? We can talk about that in the teaser?

_ = model.load_adapter(model_id, adapter_name="copilot")

if not hasattr(model, "hf_device_map"):
    model.cuda()
```

Notice that we are loading the chat assistant on top of `starcoder` instead of `starcodeplus` on which it was fine-tuned. 

Here, we will consider 2 abilities, i.e., `chatting/QA` and `code-completion` on 2 data distributions, i.e., `top 10 public hf codebase` and `generic codebase`. That gives us 4 axes on which to evaluate things. We will be carrying out qualitative analysis. 

For `chatting/QA`, the inference function is below:
```python
system_prompt = """You are a helpful, respectful and honest assistant. Always answer as helpfully \
as possible, while being safe. Your answers should not include any harmful, \
unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that \
your responses are socially unbiased and positive in nature.

If a question does not make any sense, or is not factually coherent, explain why \
instead of answering something not correct. If you don’t know the answer to a \
question, please don’t share false information."""

def get_model_pred(query, disable=False):

>     context = contextlib.nullcontext

Reviewer 3: Why is this needed? If it can be removed, please also remove it in the snippets below.

    if disable:
        context = model.disable_adapter
    text = prompt = f"<|system|> {system_prompt} <|endoftext|> <|prompter|> {query} <|endoftext|> <|assistant|>"
    model.eval()
    with context():
        outputs = model.generate(input_ids=tokenizer(text, return_tensors="pt").input_ids.cuda(), 
                                 max_new_tokens=512,
                                 temperature=0.2,
                                 top_k=50,
                                 top_p=0.95,
                                 do_sample=True,
                                 repetition_penalty=1.1,
                                 eos_token_id = tokenizer.eos_token_id)
    return tokenizer.batch_decode(outputs, skip_special_tokens=False)[0]
```

For `code-completion`, the inference function is below:

> ```

Reviewer 3: ```suggestion
```python
```

![assistant_code_hf](assets/170_personal_copilot/assistant_code_hf.png)

We can observe that the `assistant` performs similar to disabled case as it was trained on natural language conversations which didn't have any HF code repos. 

Finally, let's enable the `copilot` adapter.

![copilot_code_generic](assets/170_personal_copilot/copilot_code_generic.png)
![copilot_code_hf](assets/170_personal_copilot/copilot_code_hf.png)

We can observe that the `copilot` adapter gets it right in both case. Therefore, it performs as expected for code-completions when working with HF specific codebase as well as generic codebases.

**Now, as a user, I want to combine the ability of `assistant` as well as `copilot`. This will enable me to use it for code completion while coding in IDE and then have it as a chatbot to answer my questions regarding APIs, classes, methods, documentation. It should be able to provide answers to questions like `How do I use x`, `Please write a code snippet for Y` on my codebase.**

PEFT allows you do it by via `add_weighted_adapter`. Let's create a new adapter `code_buddy` with equal weights to `assistant` and `copilot` adapters.

> ![combining_loras](assets/170_personal_copilot/combining_loras.png)

Reviewer 3: Would this not be better as a code snippet instead of image?

model = AutoModelForCausalLM.from_pretrained(
    model, quantization_config=None, 
    device_map=None, 
    trust_remote_code=True, 
    torch_dtype=torch.bfloat16,
)

model_id = "smangrul/peft-lora-starcoder15B-v2-personal-copilot-A100-40GB-colab"
model = PeftModel.from_pretrained(model, model_id, adapter_name="copilot")

if not hasattr(model, "hf_device_map"):
    model.cuda()

# decrease the weight of loRA adapter a little bit to overcome the overfitting.
model.add_weighted_adapter(["copilot"], [0.8], "code_buddy")
model.set_adapter("code_buddy")

def get_model_pred(query, disable=False):
    context = contextlib.nullcontext
    if disable:
        context = model.disable_adapter
    text = prompt = f"Question: {query}\n\nAnswer:"
    model.eval()
    with context():
        outputs = model.generate(input_ids=tokenizer(text, return_tensors="pt").input_ids.cuda(), 
                                 max_new_tokens=1024,
                                 temperature=0.2,
                                 top_k=50,
                                 top_p=0.95,
                                 do_sample=True,
                                 repetition_penalty=1.0,

>                                  eos_token_id = tokenizer.eos_token_id)

Reviewer 4: AS before, should be formatted as code in HF.

First, let's see the output with adapter disabled to make sure it isn't part of the training data of Octocoder:

![octocoder_disabled_chat_hf](assets/170_personal_copilot/octocoder_disabled_chat_hf.png)

We can see that it fails to correctly use the API of LoraConfig or to create a PEFT model. Now, let's see it performance with the adapter enabled.

![octocoder_chat_hf](assets/170_personal_copilot/octocoder_chat_hf.png)
![octocoder_chat_generic](assets/170_personal_copilot/octocoder_chat_generic.png)

Yay! It correctly answers in detail how to create LoraConfig and related peft model along with correctly using the model name, dataset name as well as param values of LoraConfig. Also note that it does a great job at answering the generic query of using 
scrapy for crawling.
 
# How do I run it locally?

I know, after all this, you want to finetune starcoder on your codebase and use it locally on your consumer hardware such as Mac laptops with M1 GPUs, windows with RTX 4090/3090 GPUs ... 
Don't worry, we have got you covered.

> We will be using this super cool open source library [mlc-llm](https://github.com/mlc-ai/mlc-llm) 🔥. Specifically, we will be using this fork [pacman100/mlc-llm](https://github.com/pacman100/mlc-llm) which has changes to get it working with HF Code Completion extension of VS Code. On my Mac latop with M1 Metal GPU, the 15B model was painfully slow. Hence, we will go small and train a PEFT LoRA version as well as full finetuned version of `bigcode/starcoderbase-1b`. The resources are all same as above expect for the colab notebooks which are attached below:

Author: I don't think so

    "local_id": "starcoder7B-personal-copilot-merged-q4f16_1",
    "conv_template": "code_gpt",
-    "temperature": 0.7,
+    "temperature": 0.2,
-    "repetition_penalty": 1.0,
    "top_p": 0.95,
-    "mean_gen_len": 128,
+    "mean_gen_len": 64,
-    "max_gen_len": 512,
+    "max_gen_len": 64, 
    "shift_fill_factor": 0.3,
    "tokenizer_files": [
        "tokenizer.json",
        "merges.txt",
        "vocab.json"
    ],
    "model_category": "gpt_bigcode",
    "model_name": "starcoder1B-v2-personal-copilot-merged"
}
```
5. Run the local server:
```
 python -m mlc_chat.rest --model dist/starcoder1B-v2-personal-copilot-merged-q4f16_1/params --lib-path dist/starcoder1B-v2-personal-copilot-merged-q4f16_1/starcoder1B-v2-personal-copilot-merged-q4f16_1-metal.so
```
6. Change the end-point of HF Code Completion extension in VS Code to point to the local server:

![local_endpoint](assets/170_personal_copilot/local_endpoint.png)

7. Open a new file in vs code and paste the below code and have the cursor in between the doc quotes so that the model tries to infill the doc string:

![local_inference](assets/170_personal_copilot/local_inference.png)

Voila! ⭐️

> The demo at the start is this 1B model that is running locally on my Mac laptop.

Reviewer 5: How many tokens per second are you getting? I think it'd be interesting for the community as it's an usual comparison metric.

# Conclusion

> In this blog plost, we saw how to finetune `starcoder` on our personal codebase, i.e., how to create personal co-pilot. To this end, we developed 🤗 HugCoder. First, we dwelled deep into the data collection workflow. Then, we looked at comparison betgween QLoRA and Full fine-tuning. This was followed by super interesting section of combining LoRAs, transfering them which is still unexplored in text/code domain. Details about using 🤗 inference endpoint for hosting the fine-tuned model were given showing how easy it is to deploy the models. We also saw how to run these models locally to use for code completion in VS Code.

Reviewer 5: ```suggestion
In this blog plost, we saw how to finetune `starcoder` to create a personal co-pilot that knows about our code. We called it 🤗 HugCoder, as we trained it on Hugging Face code :) After looking at the data collection workflow, we compared training using QLoRA vs full fine-tuning. We also experimented by combining different LoRAs, which is still an unexplored technique in the text/code domain. For deployment, we examined remote inference using 🤗 inference endpoints, and also tried on-device completion of a smaller model with VS Code and MLC.
```

Reviewer 5: I'd maybe also add a call to action for readers to try it out.

In addition, SafeCoder also sounds like a viable approach that we could reference somewhere from this post: https://huggingface.co/blog/safecoder

Finally, a "Resources" section with all the links together in one place could also be helpful.
